### PR TITLE
doc: introduce doxyrunner extension

### PR DESCRIPTION
**Pull request by gmarull**
`doxyrunner` plugin can be used to run Doxygen build as part of the Sphinx build process. It is meant to be used with other plugins such as `breathe` in order to improve the user experience. The principal features offered by this plugin are:

- Doxygen build is managed as part of the `sphinx-build` process
- Doxyfile can be optionally preprocessed so that variables can be inserted
- Changes in the Doxygen input files are tracked so that Doxygen build is only run if necessary
- Synchronizes Doxygen XML output so that even if Doxygen is run only changed, deleted or added files are modified.

### CODE REVIEW

**Comment by mbolivar-nordic** [comment]
```
CMake Error at /home/runner/work/zephyr/zephyr/cmake/util/fmerge.cmake:17 (file):
  file failed to open for reading (No such file or directory):

    /home/runner/work/zephyr/zephyr/doc/_build/doxy.log
```

Looks like some conversions are still not being done?

Also tough to tell whether the 'Duplicate C declaration' warnings are real or just due to `-j auto`.

**Review by marc-hb — COMMENTED** [review]
>  It is meant to be used with other plugins such as breathe in order to improve the user experience.

I'm afraid I don't see yet what is the "user experience" benefit of making doxygen dependent on sphinx. Most people  use some higher level target and shouldn't notice any difference. For others who use lower level targets (for troubleshooting or performance reasons), this will make it more complicated or even impossible  to run doxygen directly, adding a layer of indirection in an area (the build) that already has too many layers of indirection.

Considering this PR increases the number of lines there should be some obvious benefits. I've done some testing and it makes no performance difference for incremental builds.

Speaking of number of lines there was BTW a never implemented suggestion to greatly simplify the timestamp script and maybe even replace it with single recursive copy python call: https://sourceforge.net/p/doxygen/mailman/message/36582383/

> Doxyfile can be optionally preprocessed so that variables can be inserted

CMake can pre-process anything, that's routine and does not require a sphinx plugin.

> The script to modify timestamps is not used anymore. It was redundant since Doxygen is only run if any of its input files changes and 

Doxygen runs really fast and running it never had any _direct_ performance impact, only an indirect one on breathe.

> in case there are no doc string changes, Doxygen still changes rendered timestamps in all XML files...

I'm very confused by this:
- The previous line just stated that Doxygen is not run anymore when there no changes
- The second line seems to miss the entire purpose of the timestamp script described in its --help and in fact the new script seems to re-implements more or less the same thing. For the record this script made incremental, non-doxygen builds orders of magnitude faster, see #13159 and also #13442

File: doc/CMakeLists.txt [hunk 1]
```diff
@@ -346,12 +308,6 @@ add_custom_target(
 )
 add_dependencies(sphinx-html content)
 
-# "breathe", the sphinx plugin that parses XML output from doxygen, has
-# an "everything on everything" dependency issue reported at:
-# https://github.com/michaeljones/breathe/issues/420 In other words
-# changing 1 source file costs the same build time than changing all
-# source files. breathe is fortunately smart enough not to run at all
-# when *nothing* has changed.
 add_custom_target(
```
[the remark this answers is not in the document]
Comment (gmarull) [reply doc/CMakeLists.txt:311]: > > It is meant to be used with other plugins such as breathe in order to improve the user experience.
> 
> I'm afraid I don't see yet what is the "user experience" benefit of making doxygen dependent on sphinx. Most people use some higher level target and shouldn't notice any difference. For others who use lower level targets (for troubleshooting or performance reasons), this will make it more complicated or even impossible to run doxygen directly, adding a layer of indirection in an area (the build) that already has too many layers of indirection.

It's not the first plugin to do this. Others such as [exhale](https://exhale.readthedocs.io/en/latest/) do this. Even `breathe` seems to support that (doesn't seem really well designed though), see https://github.com/michaeljones/breathe/blob/master/breathe/process.py

> 
> Considering this PR increases the number of lines there should be some obvious benefits. I've done some testing and it makes no performance difference for incremental builds.

I started this extension in order to simplify the Doxygen build, which now depends on a combination of CMake and an external script. It turned out that the external script is not actually needed (more details below), so it is true that we could keep using CMake for the Doxygen part. The main benefit right now is that we do not need to keep Doxygen input file list duplicate in both Doxyfile and CMakeLists.txt. The extension takes care of parsing that to then keep track of Doxygen changes.
> 
> Speaking of number of lines there was BTW a never implemented suggestion to greatly simplify the timestamp script and maybe even replace it with single recursive copy python call: https://sourceforge.net/p/doxygen/mailman/message/36582383/

The modification time script becomes redundant once input dependency tracking is introduced. With that feature in place, Doxygen is only run when any of the input file is changed. There is an edge case that the script could solve: what if I edit .c/.h, but I don't change any docstring? In theory the generated XML files should be the same, so I could skip updating the XML files given to Sphinx. Well, it turns out that Doxygen renders the timestamp to all XML files, so that case can't be fixed.

> 
> > Doxyfile can be optionally preprocessed so that variables can be inserted
> 
> CMake can pre-process anything, that's routine and does not require a sphinx plugin.

That's true. I added this facility to make transition easier (it's really 5 LOC + extension options).

> 
> > The script to modify timestamps is not used anymore. It was redundant since Doxygen is only run if any of its input files changes and
> 
> Doxygen runs really fast and running it never had any _direct_ performance impact, only an indirect one on breathe.
> 

Detailed above.

> > in case there are no doc string changes, Doxygen still changes rendered timestamps in all XML files...
> 
> I'm very confused by this:
> 
> * The previous line just stated that Doxygen is not run anymore when there no changes

And that's true because there is input dependency tracking.

> * The second line seems to miss the entire purpose of the timestamp script described in its --help and in fact the new script seems to re-implements more or less the same thing. For the record this script made incremental, non-doxygen builds orders of magnitude faster, see #13159 and also #13442

At the time #13159 was introduced CMake did not track Doxygen input file changes. This is not true anymore in both current CMake nor the `doxyrunner` extension. #13442 could have deleted the timestamp script.

Maybe I'm missing something here, but on my tests I see the expected behavior without caring about modification times at all.

**Comment by gmarull** [comment]
> > It is meant to be used with other plugins such as breathe in order to improve the user experience.
> 
> I'm afraid I don't see yet what is the "user experience" benefit of making doxygen dependent on sphinx. Most people use some higher level target and shouldn't notice any difference. For others who use lower level targets (for troubleshooting or performance reasons), this will make it more complicated or even impossible to run doxygen directly, adding a layer of indirection in an area (the build) that already has too many layers of indirection.

It's not the first plugin to do this. Others such as [exhale](https://exhale.readthedocs.io/en/latest/) do this. Even `breathe` seems to support that (doesn't seem really well designed though), see https://github.com/michaeljones/breathe/blob/master/breathe/process.py

> 
> Considering this PR increases the number of lines there should be some obvious benefits. I've done some testing and it makes no performance difference for incremental builds.

I started this extension in order to simplify the Doxygen build, which now depends on a combination of CMake and an external script. It turned out that the external script is not actually needed (more details below), so it is true that we could keep using CMake for the Doxygen part. The main benefit right now is that we do not need to keep Doxygen input file list duplicate in both Doxyfile and CMakeLists.txt. The extension takes care of parsing that to then keep track of Doxygen changes.
> 
> Speaking of number of lines there was BTW a never implemented suggestion to greatly simplify the timestamp script and maybe even replace it with single recursive copy python call: https://sourceforge.net/p/doxygen/mailman/message/36582383/

The modification time script becomes redundant once input dependency tracking is introduced. With that feature in place, Doxygen is only run when any of the input file is changed. There is an edge case that the script could solve: what if I edit .c/.h, but I don't change any docstring? In theory the generated XML files should be the same, so I could skip updating the XML files given to Sphinx. Well, it turns out that Doxygen renders the timestamp to all XML files, so that case can't be fixed.

> 
> > Doxyfile can be optionally preprocessed so that variables can be inserted
> 
> CMake can pre-process anything, that's routine and does not require a sphinx plugin.

That's true. I added this facility to make transition easier (it's really 5 LOC + extension options).

> 
> > The script to modify timestamps is not used anymore. It was redundant since Doxygen is only run if any of its input files changes and
> 
> Doxygen runs really fast and running it never had any _direct_ performance impact, only an indirect one on breathe.
> 

Detailed above.

> > in case there are no doc string changes, Doxygen still changes rendered timestamps in all XML files...
> 
> I'm very confused by this:
> 
> * The previous line just stated that Doxygen is not run anymore when there no changes

And that's true because there is input dependency tracking.

> * The second line seems to miss the entire purpose of the timestamp script described in its --help and in fact the new script seems to re-implements more or less the same thing. For the record this script made incremental, non-doxygen builds orders of magnitude faster, see #13159 and also #13442

At the time #13159 was introduced CMake did not track Doxygen input file changes. This is not true anymore in both current CMake nor the `doxyrunner` extension. #13442 could have deleted the timestamp script.

Maybe I'm missing something here, but on my tests I see the expected behavior without caring about modification times at all.

**Comment by gmarull** [comment]
> ```
> CMake Error at /home/runner/work/zephyr/zephyr/cmake/util/fmerge.cmake:17 (file):
>   file failed to open for reading (No such file or directory):
> 
>     /home/runner/work/zephyr/zephyr/doc/_build/doxy.log
> ```
> 
> Looks like some conversions are still not being done?
> 
> Also tough to tell whether the 'Duplicate C declaration' warnings are real or just due to `-j auto`.

I'll look at that part, we don't actually need to merge when using `doxyrunner`. In fact, Doxygen logs are empty nowadays because `QUIET=YES` in the Doxyfile...

**Comment by gmarull** [comment]
@marc-hb modified `doxyrunner` to handle the modification times edge case.

File: doc/CMakeLists.txt [hunk 2]
```diff
@@ -140,54 +138,11 @@ add_custom_target(
...
-# But it's not because it has a cascade effect on sphinx:
-# https://sourceforge.net/p/doxygen/mailman/message/36580807/
-# For now this optimization speeds-up non-doxygen documentation work
-# only (by one order of magnitude).
 add_custom_target(
-  doxy_real_modified_times
-  COMMAND ${CMAKE_COMMAND} -E env
-    ${PYTHON_EXECUTABLE} _scripts/restore_modification_times.py
-    --loglevel  WARN  ${DOXY_OUT}/xml
+  doxygen
```
Comment (carlescufi) [inline doc/CMakeLists.txt:142]: Why are you keeping this target at all?

**Review by carlescufi — COMMENTED** [review]
+1 for getting rid of the duplication of the doxygen input files, and for simplifying the CMake file. I also think that moving to sphinx extensions seems to be the way to go here, instead of doing complex operations in CMake.

File: doc/CMakeLists.txt (hunk 2 above)
Comment (gmarull) [reply doc/CMakeLists.txt:142]: Because Zephyr still publishes the standalone Doxygen documentation https://docs.zephyrproject.org/apidoc/latest/. I'm not sure though if `doxygen` target is used for that.

**Comment by carlescufi** [comment]
@marc-hb @mbolivar @tejlmand I think the point here to be discussed is whether we want to move some of the functionality that is currently being offered by CMake into Sphinx plugins. I certainly see some advantages:

- A single programming language, everything is written in Python
- A single framework, Sphinx takes care of everything. In the future we could go as far as removing the CMake file altogether
- CMake is currently being used as a scripting language in `doc/CMakeLists.txt`. The single exception to that is the `DEPENDS ${DOXY_SOURCES}` statement in Doxygen custom command. That makes use of CMake's ability to track changes to dependencies, and has been rewritten in Python here

In general using CMake as a scripting language is not something I believe we should do, because it's more verbose and hard to follow, and you anyways always need to delegate to Python sooner or later as tasks become more complex, so it ends up becoming a combination of languages that you need to integrate with each other. So for that reason only I think it's worth considering a transition to `sphinx-build` as our main entry point for the documentation build instead of CMake. But I do agree that "losing" the ability to track file changes in dependencies is a downside.

File: doc/CMakeLists.txt (hunk 2 above)
Comment (carlescufi) [reply doc/CMakeLists.txt:142]: OK, in that case, and if we want to be able to keep doing that, I think it's a plus point towards having CMake as a front end to Sphinx, because otherwise we are sort of duplicating the target here.

Comment (gmarull) [reply doc/CMakeLists.txt:142]: Just checked, and no, it just assumes Doxygen `html` output is located at `_build/doxygen/html`. I've updated `doxyrunner` so that it also syncs `html`, this way CI is left unaffected.

Comment (gmarull) [reply doc/CMakeLists.txt:142]: I'd maintain for now the `doxygen` target, there may be users that build the standalone Doxygen API pages.

Comment (gmarull) [reply doc/CMakeLists.txt:142]: If we want to simplify the Doxygen standalone usage further, we could also create a regular Doxyfile by:

- Replacing `@ZEPHYR_BASE@` for `../`
- Replacing `@OUT_DIR@` for some pre-defined build directory, e.g. `_build/doxygen`

And always run Doxygen from the `doc` folder.

Comment (carlescufi) [reply doc/CMakeLists.txt:142]: this last suggestion seems quite reasonable, in order to be able to run doxygen locally.

**Comment by carlescufi** [comment]
@gmarull @marc-hb @mbolivar-nordic @tejlmand 
I think that regardless of whether we use the plugin or not, we should replace the complexity in `restore_modification_times.py` with something like what the plugin does:

```
    prev_xmldir = prev_outdir / xml_output[0]
    latest_xmldir = latest_outdir / xml_output[0]
    if prev_xmldir.exists():
        dcmp = filecmp.dircmp(latest_xmldir, prev_xmldir)

        for file in dcmp.right_only:
            (Path(dcmp.right) / file).unlink()

        for file in dcmp.left_only + dcmp.diff_files:
            shutil.copy(Path(dcmp.left) / file, Path(dcmp.right) / file)

        shutil.rmtree(latest_xmldir)
    else:
        latest_xmldir.rename(prev_xmldir)
```

This would also close #13639, because no need to use that library if the code is so small.

**Comment by mbolivar-nordic** [comment]
Since @gmarull has taken over the maintenance of the documentation build system I have no objections to deferring to his judgement as long as there are no performance/correctness issues. I've personally long given up on trying to understand these scripts.

FWIW on the `-j auto` issue I think the right thing is to get rid of .known-issues entries for documentation entirely and error out on any warnings at all; I'd be shocked if the .known-issues we have aren't just papering over breathe bugs or bugs in our usage of breathe.

I do think that being able to run doxygen standalone is a critical feature. I use it personally.

**Comment by gmarull** [comment]
> FWIW on the `-j auto` issue I think the right thing is to get rid of .known-issues entries for documentation entirely and error out on any warnings at all; I'd be shocked if the .known-issues we have aren't just papering over breathe bugs or bugs in our usage of breathe.
> 
+1 on that.

**Comment by carlescufi** [comment]
@marc-hb do you want to weigh in here? @gmarull is taking the maintenance of the documentation build system and we'd like to move forward one way or the other.

File: doc/CMakeLists.txt (hunk 2 above)
Comment (tejlmand) [reply doc/CMakeLists.txt:142]: > If we want to simplify the Doxygen standalone usage further, we could also create a regular Doxyfile by:
> 
>     * Replacing `@ZEPHYR_BASE@` for `../`
> 

Actually, the CMake file still honors the ZEPHYR_BASE from the environment.
So if you start forcing Doxygen to always using `../` then doxygen might use another Zephyr base than rest of doc build that is invoked through CMake.

So if changing this, then the CMake file should be updated accordingly.

>     * Replacing `@OUT_DIR@` for some pre-defined build directory, e.g. `_build/doxygen`
> 
> 
> And always run Doxygen from the `doc` folder by just typing `doxygen`. I'd personally do that, CMake adds no value here.

When using CMake, users can specify their own build folder.
That means, I can do:
```
$ cd <path>/zephyr
$ cmake -GNinja -Bbuild_doc doc/
$ ninja -Cbuild_doc
```
so changing that will suddenly give me an extra `_build/doxygen` folder I need to consider.

**Review by marc-hb — COMMENTED** [review]
Sorry I don't have the time to review everything right now, maybe in a few days (=too late?) So I'll just share my "core" principles that seem relevant here:
- As fewer layers of indirections as possible. So, if this PR adds one layer of indirection to run doxygen: bad. If this merely replaces the CMake indirection with a sphinx indirection: OK but see below.
- Layers of indirections should be as transparent and "optional" as possible. "Optional" for the lack of a better word: I mean the ability to quickly "peel" the layer of indirection and copy/paste the lower-level command directly for investing its failures. For instance `CMake's VERBOSE=1` achieves this, it prints commands that can be most of the time re-run by merely copying them to the command line. So this PR should preserve the ability to just copy/paste the doxygen invocation on the command line to re-run doxgen alone and debug doxygen failures directly and quickly.
- I agree CMake is a terrible scripting language, Python all the way, except for... tracking dependencies of course! That's why CMake sits at the top-level. I mean it's better to have CMake invoke a sequence of smaller Python scripts that can be run independently compared to running one giant and opaque Python script that does everything in one go. Modularity and incremental builds for the win and I doubt sphinx can do that.

I quickly wrote the following Python function below that can implement the "restore timestamps" simplification suggested in https://sourceforge.net/p/doxygen/mailman/message/36582383/ and in #13639 and it seems to work. Maybe it could help simplify this PR too?
The idea is to replace the "two ways" synchronization with this simpler, one-way build:
```
C sources ---doxygen()--->  dumb doxygen output  ---recursive_copy_if_changed()---> smart doxygen output
```
```python
def rec_copy_if_changed(src, dst):

   def unchanged_files(srcdirstr, all_files):
       # closure on dst and src
       srcdir = Path(srcdirstr)
       reldir = srcdir.relative_to(src)
       dstdir = dst / reldir

       unchanged = []
       for f in all_files:
           srcf = srcdir / Path(f)
           if srcf.is_symlink() or not srcf.is_file():
               continue

           dstf = dstdir / Path(f)
           if not dstf.exists():
               continue

           if filecmp.cmp(srcf, dstf):
               unchanged.append(f)

       return unchanged

   shutil.copytree(src, dst, ignore=unchanged_files, dirs_exist_ok=True)
```
`dirsync` can probably do the same however it's yet another Python dependency, as @carlescufi said not worth it to save just a few lines.

**Comment by gmarull** [comment]
> Sorry I don't have the time to review everything right now, maybe in a few days (=too late?) So I'll just share my "core" principles that seem relevant here:
> 
> * As fewer layers of indirections as possible. So, if this PR adds one layer of indirection to run doxygen: bad. If this merely replaces the CMake indirection with a sphinx indirection: OK but see below.

To be precise it replaces one level of indirection (CMake) with a Sphinx extension. After these changes CMake becomes purely a scripting facility in regard to Sphinx, which means we could potentially remove it in the future.

> * Layers of indirections should be as transparent and "optional" as possible. "Optional" for the lack of a better word: I mean the ability to quickly "peel" the layer of indirection and copy/paste the lower-level command directly for investing its failures. For instance `CMake's VERBOSE=1` achieves this, it prints commands that can be most of the time re-run by merely copying them to the command line. So this PR should preserve the ability to just copy/paste the doxygen invocation on the command line to re-run doxgen alone and debug doxygen failures directly and quickly.

`doxygen` target is still offered, so the extension is optional for the standalone build. The extension is actually quite transparent, it offers the option to "print" the Doxygen logs (setting `doxyrunner_silent = False`) and fails if Doxygen returns non-zero.

> * I agree CMake is a terrible scripting language, Python all the way, except for... tracking dependencies of course! That's why CMake sits at the top-level. I mean it's better to have CMake invoke a sequence of smaller Python scripts that can be run independently compared to running one giant and opaque Python script that does everything in one go. Modularity and incremental builds for the win and I doubt sphinx can do that.
> 

I think Sphinx extensions comply with the modularity principle. They integrate well with the Sphinx runner and offer access to all the Sphinx facilities. In fact, I think complexity is reduced since we remove a non-trivial dependency (CMake) or duplicate Doxygen input file list.

> I quickly wrote the following Python function below that can implement the "restore timestamps" simplification suggested in https://sourceforge.net/p/doxygen/mailman/message/36582383/ and in #13639 and it seems to work. Maybe it could help simplify this PR too?

I think current `sync_doxygen` function solves the timestamp problem with ~30LOC, and anyway, it's an edge case as mentioned previously.

> The idea is to replace the "two ways" synchronization with this simpler, one-way build:
> 
> ```
> C sources ---doxygen()--->  dumb doxygen output  ---recursive_copy_if_changed()---> smart doxygen output
> ```
> 

This is already achieved with the current implementation.

> ```python
> def rec_copy_if_changed(src, dst):
> 
>    def unchanged_files(srcdirstr, all_files):
>        # closure on dst and src
>        srcdir = Path(srcdirstr)
>        reldir = srcdir.relative_to(src)
>        dstdir = dst / reldir
> 
>        unchanged = []
>        for f in all_files:
>            srcf = srcdir / Path(f)
>            if srcf.is_symlink() or not srcf.is_file():
>                continue
> 
>            dstf = dstdir / Path(f)
>            if not dstf.exists():
>                continue
> 
>            if filecmp.cmp(srcf, dstf):
>                unchanged.append(f)
> 
>        return unchanged
> 
>    shutil.copytree(src, dst, ignore=unchanged_files, dirs_exist_ok=True)
> ```
> 
> `dirsync` can probably do the same however it's yet another Python dependency, as @carlescufi said not worth it to save just a few lines.

**Comment by gmarull** [comment]
> Can this comment be moved to wherever the same bug causes the same, very severe performance issue?

done

**Review by marc-hb — COMMENTED** [review]
Thanks @gmarull for your patience making some adjustments, preserving some features and thoroughly explaining and documenting everything. We had a look at 2 years-old PRs, I think this proves the latter is useful :-)

Not going to hit the Approve button because I barely looked at the code but overall LGTM.

2 nits:
- Maybe there should be a bit more verbose comments in the code and especially with respect to performance. I mean comments copied straight from this PR discussion. Because the doc build is unfortunately a bit complicated and its performance aspects even less intuitive, I think the current discussions demonstrate that. So I suggest reviewing every explanation you gave in this PR and gauge whether it's source-worthy (if not already in a comment).
- rsync/dirsync function, see below.


> The modification time script becomes mostly redundant once input dependency tracking is introduced. With that feature in place, Doxygen is only run when any of the input file is changed.

Finally got it, thanks.

> There is an edge case that the script could solve: what if I edit .c/.h, but I don't change any docstring? In theory the generated XML files should be the same, so I could skip updating the XML files given to Sphinx. We can add this case to the extension, I wonder though if it is worth it.

I don't think it is such an edge case, I think it's easy to imagine someone multi tasking between `.rst` documentation work and C coding that does not involve doxygen changes.

We should also hope that the the "everything depends on everything" breathe bug might get fixed some day. Then smart doxygen output with correct timestamps should make a huge difference because breathe is so time consuming.

Also keep in mind developers' patience is extremely limited with respect to documentation, much more limited than with code. So every little helps!

Speaking of developer (im)patience and incremental build performance, here's one simple CI suggestion: after a succesful doc build in CI, perform a single change in a single .rst file and build again. Time the second, incremental build, fail CI if that takes more than 10 or 15 seconds. Same again with a single change to a single C file (with a higher threshold).

**Comment by galak** [comment]
dev-review (Apr 15): have a commit that uses a simple script first, than follow on commit that does the rest (@carles to add more details)

File: doc/_extensions/zephyr/doxyrunner.py [hunk 3]
```diff
@@ -0,0 +1,332 @@
...
+
+        for file in dcmp.right_only:
+            (Path(dcmp.right) / file).unlink()
+
+        for file in dcmp.left_only + dcmp.diff_files:
+            shutil.copy(Path(dcmp.left) / file, Path(dcmp.right) / file)
+
+        shutil.rmtree(latest_xmldir)
+    else:
+        latest_xmldir.rename(prev_xmldir)
```
[the remark this answers is not in the document]
Comment (gmarull) [reply doc/_extensions/zephyr/doxyrunner.py:268]: I think it's better to not anticipate future needs and stay specific with Doxygen/Sphinx needs for now. A full-featured rsync function may imply handling edge cases not needed/covered here. I agree though that we should have a common utility library if in the future other plugins need similar functionality.

File: doc/CMakeLists.txt (hunk 1 above)
[the remark this answers is not in the document]
Comment (gmarull) [reply doc/CMakeLists.txt:302]: added reference to breathe 420 issue in the doxyrunner extension.

File: doc/_extensions/zephyr/doxyrunner.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (marc-hb) [reply doc/_extensions/zephyr/doxyrunner.py:268]: I wrote "generic", not "full-featured". I'm only suggesting a small code style change, not a new "common utility library". It's much faster to read a one-line `copy_something()` function call and abstraction than infer the same despite `prev_xmldir` and other symbol noise unrelated to some pure copying logic. It may require a bit of effort https://martinfowler.com/bliki/TwoHardThings.html

No big deal.

[the remark this answers is not in the document]
Comment (gmarull) [reply doc/_extensions/zephyr/doxyrunner.py:268]: I think the function signature and its docstrings make its purpose quite clear:

```python
def sync_doxygen(doxyfile: str, outdir: Path, latest_outdir: Path) -> None:
    """Synchronize Doxygen output with a previous build.
    This function makes sure that only new, deleted or changed files are
    actually modified in the Doxygen XML output. Latest HTML content is just
    moved.
    Args:
        doxyfile: Contents of the Doxyfile.
        outdir: Doxygen output directory.
        latest_outdir: Latest Doxygen build output directory.
    """
```

I can add some more in-line comments if it helps

[the remark this answers is not in the document]
Comment (marc-hb) [reply doc/_extensions/zephyr/doxyrunner.py:268]: I think we keep talking across each other so I will be more specific with some code.

First I find the name `latest_outdir` confusing because it makes me think about source control and/or "latest and greatest" and this is neither. In fact the argument name where the function is used is "tmpdir" which sounds pretty different and much more like it. `tmp_builddir` would be better or you could also go with a simple  `sync_doxygen(doxyfile: str, src_dir: Path, dest_dir: Path)` just like `cp`, `rsync`, `memcpy`, `shutil.copy` etc. Swapping the argument order would also be consistency with all these and with your own `filecmp.dircmp` invocation.

Now back to my first suggestion:

```python

 ...

    if dst_xmldir.exists():
        smart_rcopy(src_xmldir, dst_xmldir)
        shutil.rmtree(src_xmldir)
    else:
        src_xmldir.rename(dst_xmldir)

# (Nothing about XML or doxygen in this simple and generic function,
# it's contained above.)
def smart_rcopy(src, dst):
   """Recursive copy that does not copy identical files not to update their
    timestamps"""

        dcmp = filecmp.dircmp(src, dst)

        for file in dcmp.right_only:
            (Path(dcmp.right) / file).unlink()

        for file in dcmp.left_only + dcmp.diff_files:
            shutil.copy(Path(dcmp.left) / file, Path(dcmp.right) / file)
```

[the remark this answers is not in the document]
Comment (gmarull) [reply doc/_extensions/zephyr/doxyrunner.py]: I have modified the function signature so that it has the src/dest order, but I don't see the real benefit of adding `smart_rcopy`, so I haven't changed that part.

**Comment by gmarull** [comment]
@carlescufi tested on Windows, works fine.

File: doc/_extensions/zephyr/doxyrunner.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (marc-hb) [reply doc/_extensions/zephyr/doxyrunner.py]: > but I don't see the real benefit of adding smart_rcopy

Like so many other things in programming, it's just a small abstraction to help (a little bit) with readability. So the "smart copy + delete XML" part of the logic is not obscured by distracting details about how the smart copy is implemented, and the smart copy implementation that does not care about XML does not have distracting XML symbol names. When I reviewed the code, both these distractions slowed me down enough to care and provide this suggestion. Giving the "smart copy" abstraction an (even better?) name also boosts readability.

"“Programs must be written for people to read, and only incidentally for machines to execute.”

It's a small thing, never mind.

**Comment by carlescufi** [comment]
Despite my reluctance to the move from CMake to Sphinx for the doxygen execution, my requests have been addressed. However, there's one more change I'd like to see before we merge this: An update to the [doc README](https://github.com/zephyrproject-rtos/zephyr/blob/master/doc/README.rst) that briefly describes what task in the diagram is performed where (a simple bullet list with a quick introductory paragraph should be enough).

**Review by carlescufi — APPROVED** [review]

**Review by mbolivar-nordic — APPROVED** [review]

**Comment by marc-hb** [comment]
> Speaking of developer (im)patience and incremental build performance, here's one simple CI suggestion: after a succesful doc build in CI, perform a single change in a single .rst file and build again. Time the second, incremental build, fail CI if that takes more than 10 or 15 seconds. Same again with a single change to a single C file (with a higher threshold).

No proof this was introduced by this but I found a serious performance regression, filed in #36033
